src/commands/database-remove.js

```diff
@@ -38,32 +40,199 @@ module.exports = new Command("database:remove <path>")
+              } else if (res.statusCode == 400 || res.statusCode == 413) {
+                // timeout or payload too large
+                // large subtree, recursive delete for each subtree
+                return resolve("larger");
+              } else {
+                return reject(responseToError(res, body));
+              }
+            });
+          });
+        });
+      }
+
+      function listPath(path) {
+        var url = utils.addSubdomain(api.realtimeOrigin, options.instance) + path + ".json?";
+        if (path === "/") {
+          url += querystring.stringify({
+            shallow: true,
+          });
+        } else {
+          url += querystring.stringify({
+            shallow: true,
+            limitToFirst: "100",
+          });
+        }
+        var reqOptions = {
+          url: url,
+        };
+        return api.addRequestHeaders(reqOptions).then(function(reqOptionsWithToken) {
+          return new Promise(function(resolve, reject) {
+            request.get(reqOptionsWithToken, function(err, res, body) {
+              if (err) {
+                return reject(
+                  new FirebaseError("Unexpected error while list subtrees", {
+                    exit: 2,
+                  })
+                );
+              } else if (res.statusCode >= 400) {
+                return reject(responseToError(res, body));
+              }
+              var data = {};
+              try {
+                data = JSON.parse(body);
+              } catch (e) {
+                return reject(
+                  new FirebaseError("Malformed JSON response in shallow get ", {
+                    exit: 2,
+                    original: e,
+                  })
+                );
+              }
+              if (data) {
+                var keyList = Object.keys(data);
+                return resolve(keyList);
+              } else {
+                return resolve([]);
+              }
+            });
+          });
+        });
+      }
+
+      var deleteQueue = [path];
+      var failures = [];
+      var openChunkDeleteJob = 0;
+
+      function dfsLoop() {
+        if (failures.length !== 0) {
+          return true;
+        }
+        if (deleteQueue.length === 0) {
+          return openChunkDeleteJob === 0;
+        }
+        if (openChunkDeleteJob < options.concurrent) {
+          chunckedDelete(deleteQueue.pop());
+        }
+        return false;
+      }
+
+      function chunckedDelete(path) {
+        openChunkDeleteJob += 1;
+        return prefetchTest(path)
+          .then(function(test) {
+            if (test === "small") {
+              return deletePath(path);
+            } else if (test === "larger") {
+              return listPath(path).then(function(pathList) {
+                if (pathList) {
+                  deleteQueue.push(path);
+                  for (var i = 0; i < pathList.length; i++) {
+                    var subPath = path + (path === "/" ? "" : "/") + pathList[i];
```

Comment:
can this be replace with `require("path").join`? 
![image](https://user-images.githubusercontent.com/160236/47582658-d0740e80-d909-11e8-9d4d-642273978ad1.png)